### PR TITLE
feat: test5g have two modes now

### PR DESCRIPTION
**Pull request by Tinhone**

### CODE REVIEW

**Comment by kifuan** [comment]
建议不要用01，可以再加一个token

**Comment by Tinhone** [comment]
> 建议不要用01，可以再加一个token

When test5g has 3 modes, the situation may get worse🤔
How about adding a connector "<<=" ?

**Comment by Tinhone** [comment]
I should think about how to re code, but now I don't have any better ideas. I'm sorry about this, and I hope you can understand🙏
It would be better if you could provide some suggestions💪

**Comment by kifuan** [comment]
> I should think about how to re code, but now I don't have any better ideas. I'm sorry about this, and I hope you can understand🙏
> It would be better if you could provide some suggestions💪

Because I'm writing this on my mobile phone, some code may be incorrect.

Consider add 2 tokens: MUSIC and APP.

Then rewrite the parser rule test5g, which is `TEST5G SEMICOLON` before, to `TEST5G (APP | MUSIC) SEMICOLON;`

Define new ASTs like `Test5GMusic` or `Test5GApp`. You could use the names you'd like to.

See the two modes quite differently, as two different grammar.

It might not be the best solution, but It's my first idea.

**Comment by kifuan** [comment]
If you'd like to, you can just provide a new function in speed_tester.py and I work for the lexer & parser.

**Comment by Tinhone** [comment]
> > I should think about how to re code, but now I don't have any better ideas. I'm sorry about this, and I hope you can understand🙏
> > It would be better if you could provide some suggestions💪
> 
> Because I'm writing this on my mobile phone, some code may be incorrect.
> 
> Consider add 2 tokens: MUSIC and APP.
> 
> Then rewrite the parser rule test5g, which is `TEST5G SEMICOLON;` before, to `TEST5G (APP | MUSIC) SEMICOLON;`
> 
> Define new ASTs like `Test5GMusicAST` and `Test5GAppAST`. You could use the names you'd like to.
> 
> See the two modes quite differently, as two different grammar.
> 
> It might not be the best solution, but It's my first idea.

Good idea! Thanks very much for your advice!
好主意! 非常感谢您的建议!

**Comment by kifuan** [comment]
还不睡啊，我是刚醒了

**Comment by Tinhone** [comment]
I hope the code I wrote expressed your meaning🙏

File: helang/u8.py [hunk 1]
```diff
@@ -124,6 +124,11 @@ def __eq__(self, other):
         if isinstance(other, list):
             return self == U8(other)
 
+        if isinstance(other, int):
+            if self.value[0] != other:
+                return False
+            return True
```
Comment (kifuan) [inline helang/u8.py:130]: `return self.value[0] == other`

File: helang/speed_tester.py [hunk 2]
```diff
@@ -35,20 +48,40 @@
...
-def run_speed_test():
-    print('Cyber DJ is downloading musics via 5G...')
-    random.shuffle(MUSICS)
-    total_size = 0
-    for music in MUSICS:
-        curr_size = random.randint(*MUSIC_SIZE_MB_RANGE)
-        file = music + random.choice(SUFFIXES)
-        vip_suffix = '' if music == 'ZOOD' else ' [VIP]'
+def run_speed_test(argument):
+    if argument == 0:
```
Comment (kifuan) [inline helang/speed_tester.py:52]: Split them into two functions.

File: helang/he_ast.py [hunk 3]
```diff
@@ -135,7 +135,19 @@ def evaluate(self, env: Dict[str, U8]) -> U8:
 
 class Test5GAST(AST):
     def evaluate(self, env: Dict[str, U8]) -> U8:
-        run_speed_test()
+        run_speed_test(0)
+        return U8()
+
+
+class Test5GMusicAST(AST):
+    def evaluate(self, env: Dict[str, U8]) -> U8:
```
Comment (kifuan) [inline helang/he_ast.py:143]: We don't need this AST.
It does the same thing as `Test5GAST`, isn't it?

File: README-EN.MD [hunk 4]
```diff
@@ -321,6 +321,19 @@ It's very simple, and all you need is to start a new line and type:
 
 ```c
 test5g;
+//Equivalent to the code "test5g music;"
```
Comment (kifuan) [inline README-EN.MD:324]: DON'T EDIT EN README.

Edit README.MD, and we will let particular person to do this.

File: helang/u8.py (hunk 1 above)
Comment (kifuan) [reply helang/u8.py:130]: Why do you need to do this?

It seems we doesn't need to compare with integer now.

Comment (kifuan) [reply helang/u8.py:130]: So you can just delete these statements.

**Comment by kifuan** [comment]
Please don't do so much thing in single PR.

You can open another one for other features/refactors.

**Comment by kifuan** [comment]
You could create new branches for each PR.

**Comment by Tinhone** [comment]
> Please don't do so many things in single PR.
> 
> You can open another one for other features/refactors.

I'm sorry for causing you trouble. I'll learn my lesson
很抱歉给您添麻烦。我会吸取教训的

File: helang/he_ast.py (hunk 3 above)
Comment (Tinhone) [reply helang/he_ast.py:137]: Can I move `Test5GAST` instead of `Test5GMusicAST`?
我可以移除 `Test5GAST` 而不是 `Test5GMusicAST` 吗？

File: README-EN.MD [hunk 5]
```diff
@@ -38,7 +38,6 @@ Very good works! Love from **Cyber Spaces**!
 > Nowadays in the 21st century, `Helang`, `Jvav` and `C--` have become the three irreplaceable programming languages.
 >
 > Saint He. (2202). *A Brief **He**story of Cyber Programming Languages*. Lithang Computer Science and Technology Press.
-
```
Comment (kifuan) [inline README-EN.MD:41]: Why do you remove these blank lines?

File: helang/parser.py [hunk 6]
```diff
@@ -174,24 +174,17 @@ def _root_parse_expr_statement(self) -> AST:
...
         """
-        test_5g: TEST_5G MUSIC SEMICOLON;
+        test_5g: TEST_5G (none | MUSIC) SEMICOLON;
         :return: AST for testing 5G.
         """
         self._expect(TokenKind.TEST_5G)
-        self._expect(TokenKind.T5G_MUSIC)
+        try:
+            self._expect(TokenKind.T5G_MUSIC)
+        except:
```
Comment (kifuan) [inline helang/parser.py:186]: `self._expect` will change `self._pos`, which could lead unexpectable bugs.

Please delete this `try-expect` expression and I will find a solution for this situation.

File: helang/parser.py [hunk 7]
```diff
@@ -174,24 +174,17 @@ def _root_parse_expr_statement(self) -> AST:
...
-        """
-        self._expect(TokenKind.TEST_5G)
-        self._expect(TokenKind.SEMICOLON)
-        return Test5GAST()
-
     @_ruled_methods.bind(Rule.ROOT)
     def _root_parse_test_5g(self) -> Test5GMusicAST:
         """
-        test_5g: TEST_5G MUSIC SEMICOLON;
+        test_5g: TEST_5G (none | MUSIC) SEMICOLON;
```
Comment (kifuan) [inline helang/parser.py:180]: As I mentioned below, delete this none case.

File: helang/parser.py (hunk 6 above)
Comment (kifuan) [reply helang/parser.py:186]: So you should also delete none| in the docstring.

File: README-EN.MD (hunk 5 above)
Comment (Tinhone) [reply README-EN.MD:41]: It's my carelessness. I'm sorry
是我的疏忽，对不起

**Comment by Kerushii** [comment]
> > Please don't do so many things in single PR.
> > You can open another one for other features/refactors.
> 
> I'm sorry for causing you trouble. I'll learn my lesson 很抱歉给您添麻烦。我会吸取教训的

1050458482 加qq 群
h

**Comment by kifuan** [comment]
我不就是没自己建群吗😤

**Comment by Kerushii** [comment]
不行
只能用云猫猫群

**Comment by kifuan** [comment]
给你云跌引流是吧😤

**Pull request merged**